taraxa/state/dpos/precompiled/dpos_contract.go

```diff
@@ -0,0 +1,1022 @@
+	}
+
+	state.Count++
+	self.state_put(&state_k, state)
+	self.validators.ModifyValidator(&args.Validator, validator)
+	return nil
+}
+
+func (self *Contract) undelegate(ctx vm.CallFrame, block types.BlockNum, args UndelegateArgs) error {
+	if self.undelegations.UndelegationExists(ctx.CallerAccount.Address(), &args.Validator) {
+		return ErrExistentUndelegation
+	}
+
+	validator := self.validators.GetValidator(&args.Validator)
+	if validator == nil {
+		return ErrNonExistentValidator
+	}
+	was_eligible := validator.TotalStake.Cmp(self.cfg.EligibilityBalanceThreshold) >= 0
+	prev_vote_count := vote_count(validator.TotalStake, self.cfg.EligibilityBalanceThreshold, self.cfg.VoteEligibilityBalanceStep)
+
+	delegation := self.delegations.GetDelegation(ctx.CallerAccount.Address(), &args.Validator)
+	if delegation == nil {
+		return ErrNonExistentDelegation
+	}
+
+	if delegation.Stake.Cmp(args.Amount) == -1 {
+		return ErrInsufficientDelegation
+	}
+
+	state, state_k := self.state_get(args.Validator[:], BlockToBytes(block))
+	if state == nil {
+		old_state := self.state_get_and_decrement(args.Validator[:], BlockToBytes(validator.LastUpdated))
+		state = new(State)
+		state.RwardsPer1Stake = bigutil.Add(old_state.RwardsPer1Stake, bigutil.Div(validator.RewardsPool, validator.TotalStake))
+		validator.RewardsPool = bigutil.Big0
+		validator.LastUpdated = block
+		state.Count++
+	}
+
+	// We need to claim rewards first
+	old_state := self.state_get_and_decrement(args.Validator[:], BlockToBytes(delegation.LastUpdated))
+	reward := bigutil.Sub(state.RwardsPer1Stake, old_state.RwardsPer1Stake)
+	// Reward needs to be add to callers accounts as only stake is locked
+	ctx.CallerAccount.AddBalance(bigutil.Mul(reward, delegation.Stake))
+
+	// Creating undelegation request
+	self.undelegations.CreateUndelegation(ctx.CallerAccount.Address(), &args.Validator, block+self.cfg.WithdrawalDelay, args.Amount)
+	delegation.Stake = bigutil.Sub(delegation.Stake, args.Amount)
+	validator.TotalStake = bigutil.Sub(validator.TotalStake, args.Amount)
+
+	if delegation.Stake.Cmp(bigutil.Big0) == 0 {
+		self.delegations.RemoveDelegation(ctx.CallerAccount.Address(), &args.Validator)
+	} else {
+		delegation.LastUpdated = block
+		state.Count++
+		self.delegations.ModifyDelegation(ctx.CallerAccount.Address(), &args.Validator, delegation)
+	}
+
+	self.amount_delegated = bigutil.Sub(self.amount_delegated, args.Amount)
+	eligible_now := validator.TotalStake.Cmp(self.cfg.EligibilityBalanceThreshold) >= 0
+	if was_eligible && !eligible_now {
+		self.eligible_count--
+	}
+	new_vote_count := vote_count(validator.TotalStake, self.cfg.EligibilityBalanceThreshold, self.cfg.VoteEligibilityBalanceStep)
+	if prev_vote_count != new_vote_count {
+		self.eligible_vote_count -= prev_vote_count
+		self.eligible_vote_count = Add64p(self.eligible_vote_count, new_vote_count)
+	}
+
+	// TODO: cant do this, RewardsPool as well as CommissionRewardsPool must be == 0 too
```

Comment:
Is this solved ? Especially CommissionRewardsPool seems like you can delete validator object without claiming commission rewards